cmd/queue/main.go

```diff
@@ -298,39 +297,39 @@ func main() {
 		fmt.Sprintf(":%d", v1alpha1.RequestQueuePort),
 		queue.TimeToFirstByteTimeoutHandler(http.HandlerFunc(handler), time.Duration(revisionTimeoutSeconds)*time.Second, "request timeout"))
 
-	// An `ErrServerClosed` should not trigger an early exit of
-	// the errgroup below.
-	catchServerError := func(runner func() error) func() error {
-		return func() error {
-			err := runner()
-			if err != http.ErrServerClosed {
-				return err
-			}
-			return nil
+	errChan := make(chan error, 2)
+	// Runs a server created by creator and sends fatal errors to the errChan.
+	// Does not act on the ErrServerClosed error since that indicates we're
+	// already shutting everything down.
+	catchServerError := func(creator func() error) {
+		if err := creator(); err != nil && err != http.ErrServerClosed {
+			errChan <- err
 		}
 	}
 
-	var g errgroup.Group
-	g.Go(catchServerError(server.ListenAndServe))
-	g.Go(catchServerError(adminServer.ListenAndServe))
-	g.Go(func() error {
-		<-signals.SetupSignalHandler()
-		return errors.New("Received SIGTERM")
-	})
-
-	if err := g.Wait(); err != nil {
-		logger.Errorw("Shutting down", zap.Error(err))
-	}
-
-	// Calling server.Shutdown() allows pending requests to
-	// complete, while no new work is accepted.
-	if err := adminServer.Shutdown(context.Background()); err != nil {
-		logger.Errorw("Failed to shutdown admin-server", zap.Error(err))
-	}
+	go catchServerError(server.ListenAndServe)
+	go catchServerError(adminServer.ListenAndServe)
+
+	// Blocks until we actually receive a TERM signal or one of the servers
+	// exit unexpectedly. We fold both signals together because we only want
+	// to act on the first of those to reach here.
+	select {
+	case err := <-errChan:
+		logger.Errorw("Failed to bring up queue-proxy, shutting down.", zap.Error(err))
+		os.Exit(1)
```

Comment:
1. logger.Fatalw will accomplish both lines.
2.  if we want clean exit, we can have a 
```
func main() {
  retC := 0
  defer func() { os.Exit(retC) }
...
...
   logger.Errorw("Failed...")
   retC = 1
}
```
Since os.Exit is the first in the stack will be the last to execute, ensuring all the other deferred calls have been executed.

Reply:
I've been thinking about cooking up a "ShutdownManager" library for all our components to use, that takes care of all of this. It seems quite difficult to nail graceful shutdown just right.